src/PerfCounters.cc

```diff
@@ -1090,4 +1097,96 @@ Ticks PerfCounters::read_ticks(Task* t, Error* error) {
   return ret;
 }
 
+#ifdef USEBPF
+bool PerfCounters::accelerate_async_signal(const Registers& regs) {
+  static int initialized;
+  static struct perf_event_attr attr;
+  static int bpf_prog_fd;
+  static struct user_regs_struct* bpf_regs;
+
+  if (!fd_async_signal_accelerator.is_open()) {
+    if (!initialized) {
```

Comment:
How about moving this BPF initialization code into its own function?

Reply:
It feel a bit ugly to be mashing the BPF program's global state in this function. And it's ugly to be mmapping that buffer and then leaking it to the global variable.

How hard would it be to put the BPF program and its state into its own class with proper ownership, and have each ReplaySession hold a shared pointer to an object of that class?

Reply:
Alright I reorganized this along those lines. The bpf singleton stuff lives in a BpfAccelerator class that's shared between the different PerfCounters instances.